// @bokeh/bokehjs/build/js/lib/models/renderers/renderer.js
import { View } from "../../core/view";
import * as visuals from "../../core/visuals";
import { RenderLevel } from "../../core/enums";
import { isString } from "../../core/util/types";
import { Model } from "../../model";
import { assert } from "../../core/util/assert";
import { CoordinateTransform, CoordinateMapping } from "../coordinates/coordinate_mapping";
import { BBox } from "../../core/util/bbox";
export class RendererGroup extends Model {
    static __name__ = "RendererGroup";
    constructor(attrs) {
        super(attrs);
    }
    static {
        this.define(({ Boolean }) => ({
            visible: [Boolean, true],
        }));
    }
}
export class RendererView extends View {
    static __name__ = "RendererView";
    visuals;
    _coordinates;
    get coordinates() {
        const { _coordinates } = this;
        if (_coordinates != null)
            return _coordinates;
        else
            return this._coordinates = this._initialize_coordinates();
    }
    _custom_coordinates = null;
    set coordinates(custom_coordinates) {
        this._custom_coordinates = custom_coordinates;
    }
    initialize() {
        super.initialize();
        this.visuals = new visuals.Visuals(this);
    }
    connect_signals() {
        super.connect_signals();
        const { group } = this.model;
        if (group != null) {
            this.on_change(group.properties.visible, () => {
                this.model.visible = group.visible;
            });
        }
        const { x_range_name, y_range_name } = this.model.properties;
        this.on_change([x_range_name, y_range_name], () => delete this._coordinates);
        this.connect(this.plot_view.frame.change, () => delete this._coordinates);
    }
    _initialize_coordinates() {
        if (this._custom_coordinates != null) {
            return this._custom_coordinates;
        }
        const { coordinates } = this.model;
        const { frame } = this.plot_view;
        if (coordinates != null) {
            return coordinates.get_transform(frame);
        }
        else {
            const { x_range_name, y_range_name } = this.model;
            const x_scale = frame.x_scales.get(x_range_name);
            const y_scale = frame.y_scales.get(y_range_name);
            assert(x_scale != null, `missing '${x_range_name}' range`);
            assert(y_scale != null, `missing '${y_range_name}' range`);
            return new CoordinateTransform(x_scale, y_scale);
        }
    }
    get plot_view() {
        return this.parent;
    }
    get plot_model() {
        return this.parent.model;
    }
    get layer() {
        const { overlays, primary } = this.canvas;
        return this.model.level == "overlay" ? overlays : primary;
    }
    get canvas() {
        return this.plot_view.canvas_view;
    }
    request_render() {
        this.request_paint();
    }
    request_paint() {
        this.plot_view.request_paint(this);
    }
    request_layout() {
        this.plot_view.request_layout();
    }
    notify_finished() {
        this.plot_view.notify_finished();
    }
    notify_finished_after_paint() {
        this.plot_view.notify_finished_after_paint();
    }
    get needs_clip() {
        return false;
    }
    get has_webgl() {
        return false;
    }
    /*
    get visible(): boolean {
      const {visible, group} = this.model
      return !visible ? false : (group?.visible ?? true)
    }
    */
    get displayed() {
        return this.model.visible;
    }
    render() {
        if (this.displayed) {
            this._render();
        }
        this._has_finished = true;
    }
    renderer_view(_renderer) {
        return undefined;
    }
    /**
     * Geometry setup that doesn't change between paints.
     */
    update_geometry() { }
    /**
     * Geometry setup that changes between paints.
     */
    compute_geometry() { }
    /**
     * Compute screen coordinates for a symbolic node.
     */
    resolve_node(node) {
        const target = (() => {
            if (isString(node.target)) {
                switch (node.target) {
                    case "canvas": return this.plot_view.canvas;
                    case "frame": return this.plot_view.frame;
                    case "plot": return this.plot_view;
                    case "parent": return this.parent;
                }
            }
            else {
                if (node.target instanceof Renderer) {
                    const view = this.plot_view.renderer_view(node.target);
                    if (view != null) {
                        return view;
                    }
                }
                return null;
            }
        })();
        function xy(x, y) {
            const { offset } = node;
            return { x: x + offset, y: y + offset };
        }
        if (target == null) {
            return xy(NaN, NaN);
        }
        if (!("bbox" in target && target.bbox instanceof BBox)) {
            return xy(NaN, NaN);
        }
        const { bbox } = target;
        switch (node.symbol) {
            case "top_left": return xy(bbox.left, bbox.top);
            case "top_center": return xy(bbox.hcenter, bbox.top);
            case "top_right": return xy(bbox.right, bbox.top);
            case "center_left": return xy(bbox.left, bbox.vcenter);
            case "center_center": return xy(bbox.hcenter, bbox.vcenter);
            case "center_right": return xy(bbox.right, bbox.vcenter);
            case "bottom_left": return xy(bbox.left, bbox.bottom);
            case "bottom_center": return xy(bbox.hcenter, bbox.bottom);
            case "bottom_right": return xy(bbox.right, bbox.bottom);
            case "center": return xy(bbox.hcenter, bbox.vcenter);
            case "top": return xy(NaN, bbox.top);
            case "left": return xy(bbox.left, NaN);
            case "right": return xy(bbox.right, NaN);
            case "bottom": return xy(NaN, bbox.bottom);
            default: return xy(NaN, NaN);
        }
    }
}
export class Renderer extends Model {
    static __name__ = "Renderer";
    constructor(attrs) {
        super(attrs);
    }
    static {
        this.define(({ Boolean, String, Ref, Nullable }) => ({
            group: [Nullable(Ref(RendererGroup)), null],
            level: [RenderLevel, "image"],
            visible: [Boolean, true],
            x_range_name: [String, "default"],
            y_range_name: [String, "default"],
            coordinates: [Nullable(Ref(CoordinateMapping)), null],
            propagate_hover: [Boolean, false],
        }));
    }
}
